src/librbd/ManagedLock.h

```diff
@@ -0,0 +1,184 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_MANAGED_LOCK_H
+#define CEPH_LIBRBD_MANAGED_LOCK_H
+
+#include "include/int_types.h"
+#include "include/Context.h"
+#include "include/rados/librados.hpp"
+#include "cls/lock/cls_lock_types.h"
+#include "common/Mutex.h"
+#include <list>
+#include <string>
+#include <utility>
+
+class ContextWQ;
+
+namespace librbd {
+
+template <typename Watcher>
+class ManagedLock {
+public:
+  static const std::string WATCHER_LOCK_TAG;
+
+  static ManagedLock *create(librados::IoCtx& ioctx, ContextWQ *work_queue,
+                             const std::string& oid, Watcher *watcher) {
+    return new ManagedLock(ioctx, work_queue, oid, watcher);
+  }
+
+  ManagedLock(librados::IoCtx& ioctx, ContextWQ *work_queue,
+              const std::string& oid, Watcher *watcher);
+  ~ManagedLock();
+
+  bool is_lock_owner() const;
+
+  void shut_down(Context *on_shutdown);
+
+  void acquire_lock(Context *on_acquired);
+  void release_lock(Context *on_released);
+
+  void reacquire_lock(Context *on_reacquired = nullptr);
+
+  void assert_locked(librados::ObjectWriteOperation *op, ClsLockType type);
+
+  inline librados::IoCtx& io_ctx() const {
```

Comment:
I'd vote to yank these accessors and instead directly pass the necessary values to the state machine instances -- which can still be templated on an (unused) ```ImageCtx``` instead of ```ManagedLock<DerivedLock>```